### PR TITLE
Fix integration point transformation from boundary element to face element

### LINKED ISSUE #3938: Incorrect `be_to_bfe` result for boundary element to face integration point mapping
When using `GridFunction::GetValue` or `GetVectorValue` on a boundary element for a discontinuous solution, the integration point is mapped from the boundary element to the attached face, and then to the neighboring domain element to interpolate the solution. The boundary element to face mapping is performed by `be_to_bfe` in `gridfunc.cpp`. I noticed an issue using this method for boundary elements where the result was confusingly permuted unexpectedly between the boundary element and face element. It turns out that this is resolved by accounting for all 6 possible orientations (in the case of a triangle) instead of just the current logic.

That is, modifying:

```
if (o == 2)
{
   fip.x = 1.0 - ip.x - ip.y;
   fip.y = ip.x;
}
else if (o == 4)
{
   fip.x = ip.y;
   fip.y = 1.0 - ip.x - ip.y;
}
else
{
   fip.x = ip.x;
   fip.y = ip.y;
}
```

to:

```
if (o == 0)  // 0, 1, 2
{
   fip.x = ip.x;
   fip.y = ip.y;
}
else if (o == 5)  // 0, 2, 1
{
   fip.x = ip.y;
   fip.y = ip.x;
}
else if (o == 2)  // 1, 2, 0
{
   // Same as before:
   fip.x = 1.0 - ip.x - ip.y;
   fip.y = ip.x;
}
else if (o == 1)  // 1, 0, 2
{
   fip.x = 1.0 - ip.x - ip.y;
   fip.y = ip.y
}
else if (o == 4)  // 2, 0, 1
{
   // Same as before:
   fip.x = ip.y;
   fip.y = 1.0 - ip.x - ip.y;
}
else if (o == 3)  // 2, 1, 0
{
   fip.x = ip.x;
   fip.y = 1.0 - ip.x - ip.y;
}
```

based on the orientation definitions in [`Mesh::GetTriOrientation`](https://github.com/mfem/mfem/blob/master/mesh/mesh.cpp#L5862).

This bit of code seems to date back to https://github.com/mfem/mfem/pull/1194. @mlstowell, as the author there, do you have an explanation of why the original code only accounted for `o = 2` and `o = 4`, but not other orientations? It seems like they would always be required for the vertex permutations. Does the new approach seem correct? I imagine a similar change might need to be made for quads.

If this is indeed the right approach, then I can make a PR. Probably moving this method to be a `static` member of `Mesh` would make sense instead of having it hidden inside of `gridfunc.cpp` since it depends on the orientation definitions. There is probably a good way to test this with a unit test as well that should be part of that PR.

### PR DESCRIPTION
**Pull request by sebastiangrimberg**
Resolves #3938

Replace the `be_to_bfe` method in `gridfunc.cpp` with a `static` member of the `Mesh` class which correctly accounts for all boundary element to face orientations (3D). Previously only orientations 0, 2, 4 and 0, 2, 4, 6 were accounted for for triangles and quadrilaterals, respectively. The previous behavior was not necessarily a "bug", because as used in `GridFunction` the boundary element is always "external" (single sided) and thus the face and boundary element have aligned normals by construction, which means the possible relative orientations between the two is only half the total (and these were all accounted for). The issue comes when trying to use the same approach for a boundary and face element which is internal to the computational domain, in which case all orientations need to be accounted for.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3943](https://github.com/mfem/mfem/pull/3943) | @sebastiangrimberg | @tzanio | @vladotomov + @v-dobrev + @dylan-copeland + @mlstowell | 10/20/23 | 10/28/23 | 10/31/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: mesh/mesh.hpp [hunk 1]
```diff
@@ -1409,6 +1409,12 @@ class Mesh
    /// should not be deleted by the caller.
    static FiniteElement *GetTransformationFEforElementType(Element::Type);
 
+   /** @brief For the face of a boundary element with the orientation @a o (3D),
```
Comment (vladotomov) [inline mesh/mesh.hpp:1412]: Could you mention that the function handles both internal and external boundaries?

File: mesh/mesh.hpp [hunk 2]
```diff
@@ -1409,6 +1409,12 @@ class Mesh
...
+       return the transformation of the boundary element integration point @ ip
+       to the face element. */
+   static IntegrationPoint TransformBdrElementFace(Geometry::Type geom, int o,
```
Comment (vladotomov) [inline mesh/mesh.hpp:1415]: I recommend to keep the `To` in the name like `TransformBdrElementToFace` or `TransformBdrElementToFaceElement`.

**Review by vladotomov — APPROVED** [review]

File: mesh/mesh.hpp (hunk 2 above)
Comment (sebastiangrimberg) [reply mesh/mesh.hpp]: Agreed. Done in [5cec99a](https://github.com/mfem/mfem/pull/3943/commits/5cec99aaf62e2f8273fd5de8953ee7a8d57ae648).

**Review by dylan-copeland — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
It may be worth adding support for 2D meshes in `Mesh::TransformBdrElementToFace` or creating a separate method for that. In 2D, interior boundary elements can have opposite orientation compared to the respective face. Note that in 2D, `Mesh::GetBdrElementEdges` returns orientations `+1` or `-1`. What do you think?

File: mesh/mesh.cpp [hunk 3]
```diff
@@ -6271,6 +6271,99 @@ int Mesh::CheckBdrElementOrientation(bool fix_it)
    return wo;
 }
 
+IntegrationPoint Mesh::TransformBdrElementToFace(Geometry::Type geom, int o,
+                                                 const IntegrationPoint &ip)
+{
+   IntegrationPoint fip = {};
+   if (geom == Geometry::TRIANGLE)
+   {
+      MFEM_ASSERT(o < 6, "Invalid orientation for Geometry::TRIANGLE!");
```
Comment (v-dobrev) [inline mesh/mesh.cpp:6280]: Let's check `o >= 0 && o < 6` here.

File: mesh/mesh.cpp [hunk 4]
```diff
@@ -6271,6 +6271,99 @@ int Mesh::CheckBdrElementOrientation(bool fix_it)
...
+      }
+      else if (o == 3)  // 2, 1, 0
+      {
+         fip.x = ip.x;
+         fip.y = 1.0 - ip.x - ip.y;
+      }
+   }
+   else if (geom == Geometry::SQUARE)
+   {
+      MFEM_ASSERT(o < 8, "Invalid orientation for Geometry::SQUARE!");
```
Comment (v-dobrev) [inline mesh/mesh.cpp:6314]: Let's check `o >= 0 && o < 8` here.

File: mesh/mesh.cpp [hunk 5]
```diff
@@ -6271,6 +6271,99 @@ int Mesh::CheckBdrElementOrientation(bool fix_it)
...
+      }
+      else if (o == 7)  // 3, 2, 1, 0
+      {
+         fip.x = ip.x;
+         fip.y = 1.0 - ip.y;
+      }
+   }
+   else
+   {
+      fip.x = ip.x;
```
Comment (v-dobrev) [inline mesh/mesh.cpp:6358]: If we only support 3D in this method, this should probably trigger an error.

**Comment by sebastiangrimberg** [comment]
> It may be worth adding support for 2D meshes in Mesh::TransformBdrElementToFace or creating a separate method for that. In 2D, interior boundary elements can have opposite orientation compared to the respective face. Note that in 2D, Mesh::GetBdrElementEdges returns orientations +1 or -1. What do you think?

This sounds good, done in [5bba155](https://github.com/mfem/mfem/pull/3943/commits/5bba1554771ad6c2f5b9c873cfb93cd55fae412b). Likewise the method now triggers an error as requested for an unsupported face geometry type in 2D or 3D.

**Comment by v-dobrev** [comment]
Can you please also update the doxygen for `Mesh::TransformBdrElementToFace` that it also supports 2D?

File: mesh/mesh.hpp [hunk 6]
```diff
@@ -1409,6 +1411,13 @@ class Mesh
    /// should not be deleted by the caller.
    static FiniteElement *GetTransformationFEforElementType(Element::Type);
 
+   /** @brief For the edge or face of a boundary element with the orientation
+       @a o (2D or 3D), return the transformation of the boundary element
+       integration point @ ip to the face element. Supports both internal and
+       external boundaries. */
```
Comment (v-dobrev) [inline mesh/mesh.hpp:1417]: Update doxygen that this now works in 1D as well.

It's also a good idea to add a note that in 2D this method uses orientations 0/1 (as returned by `GetBdrElementFace`), and not +1/-1.

File: mesh/mesh.hpp [hunk 7]
```diff
@@ -1330,7 +1330,9 @@ class Mesh
        element @a elem, including @a elem. */
    Array<int> FindFaceNeighbors(const int elem) const;
 
-   /// Return the index and the orientation of the face of bdr element i. (3D)
+   /** Return the index and the orientation of the vertex of bdr element i. (1D)
+       Return the index and the orientation of the edge of bdr element i. (2D)
+       Return the index and the orientation of the face of bdr element i. (3D) */
    void GetBdrElementFace(int i, int *f, int *o) const;
```
Comment (v-dobrev) [inline mesh/mesh.hpp:1336]: We should probably note that in 2D this method returns orientations 0/1, and not +1/-1 (as does `GetBdrElementEdges`).

File: mesh/mesh.hpp (hunk 6 above)
Comment (sebastiangrimberg) [reply mesh/mesh.hpp]: [2d83759](https://github.com/mfem/mfem/pull/3943/commits/2d83759e49f74d261ffeb41091f6db24111fe34c)

File: mesh/mesh.hpp (hunk 7 above)
Comment (sebastiangrimberg) [reply mesh/mesh.hpp]: [2d83759](https://github.com/mfem/mfem/pull/3943/commits/2d83759e49f74d261ffeb41091f6db24111fe34c)

**Comment by v-dobrev** [comment]
I tested (successfully) for correctness with the following code:
```c++
#include "mfem.hpp"
#include <iostream>

using namespace std;
using namespace mfem;


template <Geometry::Type geom_t, Geometry::Type face_t>
void test()
{
   Mesh mesh;
   constexpr int n1d = 1;
   switch (geom_t)
   {
      case Geometry::SEGMENT:
         mesh = Mesh::MakeCartesian1D(n1d, Element::SEGMENT);
         break;
      case Geometry::TRIANGLE:
         mesh = Mesh::MakeCartesian2D(n1d, n1d, Element::TRIANGLE);
         break;
      case Geometry::SQUARE:
         mesh = Mesh::MakeCartesian2D(n1d, n1d, Element::QUADRILATERAL);
         break;
      case Geometry::TETRAHEDRON:
         mesh = Mesh::MakeCartesian3D(n1d, n1d, n1d, Element::TETRAHEDRON);
         break;
      case Geometry::CUBE:
         mesh = Mesh::MakeCartesian3D(n1d, n1d, n1d, Element::HEXAHEDRON);
         break;
      default:
         MFEM_ABORT("");
   }
   Element *be0 = mesh.GetBdrElement(0);
   MFEM_VERIFY(be0->GetGeometryType() == face_t, "");
   int f, o;
   mesh.GetBdrElementFace(0, &f, &o);
   const Element *face = mesh.GetFace(f);
   int *be0_v = be0->GetVertices();
   const int *face_v = face->GetVertices();
   typedef Geometry::Constants<face_t> face_t_consts;
   cout << endl
        << "Element type: " << Geometry::Name[geom_t] << endl;
   cout << "Boundary/face type: " << Geometry::Name[face_t] << endl;
   for (o = 0; o < face_t_consts::NumOrient; o++)
   {
      const int *face_perm = face_t_consts::Orient[o];
      for (int i = 0; i < face_t_consts::NumVert; i++)
      {
         be0_v[i] = face_v[face_perm[i]];
      }
      IsoparametricTransformation bdr_tr, face_tr;
      mesh.GetBdrElementTransformation(0, &bdr_tr);
      mesh.GetFaceTransformation(f, &face_tr);
      IntegrationPoint bdr_ip;
      bdr_ip.Set3(0.1, 0.3, 0.0);
      int inv_o;
      mesh.GetBdrElementFace(0, &f, &inv_o);
      MFEM_VERIFY(inv_o == face_t_consts::InvOrient[o], "");
      IntegrationPoint face_ip = Mesh::TransformBdrElementToFace(
                                    be0->GetGeometryType(), inv_o, bdr_ip);
      Vector bdr_pt, face_pt;
      bdr_tr.Transform(bdr_ip, bdr_pt);
      face_tr.Transform(face_ip, face_pt);
      cout << "orientation " << o << " (inv orientation " << inv_o << "): "
           << bdr_pt.DistanceTo(face_pt) << endl;
   }
}

int main()
{
   test<Geometry::SEGMENT,     Geometry::POINT>();
   test<Geometry::TRIANGLE,    Geometry::SEGMENT>();
   test<Geometry::SQUARE,      Geometry::SEGMENT>();
   test<Geometry::TETRAHEDRON, Geometry::TRIANGLE>();
   test<Geometry::CUBE,        Geometry::SQUARE>();
   return 0;
}
```
@sebastiangrimberg, if you think it will be useful, maybe this can be added as a unit test -- I'm fine either way.

**Review by v-dobrev — APPROVED** [review]

**Comment by sebastiangrimberg** [comment]
Thanks @v-dobrev, that's a great test! Added in [026c42f](https://github.com/mfem/mfem/pull/3943/commits/026c42fc3cce959c506cdd7a80d68b71cd7afa9c).

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**